# Loading Issuance By Nationality
Loops through all of the excel files by nationality and puts them into a dataframe

In [3]:
import os
import pandas as pd

currentDirectory = os.getcwd()
initialNationalityDataframe = []

for filename in os.listdir(currentDirectory + "/NIV Issuance FY2023 Q4/By Nationality/"):
    filePath = os.path.join(currentDirectory + "/NIV Issuance FY2023 Q4/By Nationality/", filename)
    print(filename)
    month = filename.split()[0]
    year = filename.split()[1]
    df = pd.read_excel(filePath, skiprows=1)
    df['Month'] = month
    df['Year'] = year
    initialNationalityDataframe.append(df)

nationalityDataFrame = pd.concat(initialNationalityDataframe, ignore_index=True)
nationalityDataFrame

JULY 2023 - NIV Issuances by Nationality and Visa Class.xlsx
JUNE 2023 - NIV Issuances by Nationality and Visa Class.xlsx
SEPTEMBER 2023 - NIV Issuances by Nationality and Visa Class.xlsx


Nationality Visa Class  \
0                                            Afghanistan      B1/B2   
1                                            Afghanistan         F1   
2                                            Afghanistan         F2   
3                                            Afghanistan         G4   
4                                            Afghanistan        H1B   
...                                                  ...        ...   
11392                                           Zimbabwe         P1   
11393                                           Zimbabwe         Q1   
11394                                           Zimbabwe         R1   
11395                                        GRAND TOTAL        NaN   
11396  *"Non-nationality based issuances" includes in...        NaN   

       Issuances      Month  Year  
0          124.0       JULY  2023  
1           37.0       JULY  2023  
2            4.0       JULY  2023  
3           15.0       JULY  2023  
4            5.0       JULY  2023  
...          ...        ...   ...  
11392        2.0  SEPTEMBER  2023  
11393        4.0  SEPTEMBER  2023  
11394        3.0  SEPTEMBER  2023  
11395   849400.0  SEPTEMBER  2023  
11396        NaN  SEPTEMBER  2023  

[11397 rows x 5 columns]

# Loading Issuance By Post
Loops through all of the excel files by post and puts them into a dataframe

In [4]:
import os
import pandas as pd

currentDirectory = os.getcwd()
initialPostDataframe = []

for filename in os.listdir(currentDirectory + "/NIV Issuance FY2023 Q4/By Post/"):
    filePath = os.path.join(currentDirectory + "/NIV Issuance FY2023 Q4/By Post/", filename)
    print(filename)
    month = filename.split()[0]
    year = filename.split()[1]
    df = pd.read_excel(filePath, skiprows=1)
    df['Month'] = month
    df['Year'] = year
    initialPostDataframe.append(df)

postDataFrame = pd.concat(initialPostDataframe, ignore_index=True)
postDataFrame

JULY 2023 - NIV Issuances by Post and Visa Class.xlsx
JUNE 2023 - NIV Issuances by Post and Visa Class.xlsx
SEPTEMBER 2023 - NIV Issuances by Post and Visa Class.xlsx


Post Visa Class  Issuances      Month  Year
0          Abidjan         A2         16       JULY  2023
1          Abidjan      B1/B2        538       JULY  2023
2          Abidjan         C1          5       JULY  2023
3          Abidjan       C1/D          7       JULY  2023
4          Abidjan         C2          1       JULY  2023
...            ...        ...        ...        ...   ...
13413       Zagreb         O3          1  SEPTEMBER  2023
13414       Zagreb         P3          1  SEPTEMBER  2023
13415       Zagreb         R1          2  SEPTEMBER  2023
13416       Zagreb         R2          2  SEPTEMBER  2023
13417  GRAND TOTAL        NaN     849400  SEPTEMBER  2023

[13418 rows x 5 columns]

# Filter Dataframes to only look at Visa Class H1B

In [15]:
h1bPostDataFrame = postDataFrame[postDataFrame['Visa Class'] == 'H1B']

h1bPostDataFrame.to_excel('H1B_Files/H1B_By_Post_Q4_FY2023.xlsx', index=False, engine='openpyxl')


h1bPostDataFrame

Post Visa Class  Issuances      Month  Year
10         Abidjan        H1B          7       JULY  2023
31       Abu Dhabi        H1B         35       JULY  2023
57           Abuja        H1B          6       JULY  2023
77           Accra        H1B         18       JULY  2023
105    Addis Ababa        H1B          5       JULY  2023
...            ...        ...        ...        ...   ...
13303       Warsaw        H1B         19  SEPTEMBER  2023
13341        Wuhan        H1B         37  SEPTEMBER  2023
13356      Yaounde        H1B          3  SEPTEMBER  2023
13375      Yerevan        H1B          5  SEPTEMBER  2023
13402       Zagreb        H1B          7  SEPTEMBER  2023

[533 rows x 5 columns]

In [16]:
h1bNationalityDataFrame= nationalityDataFrame[nationalityDataFrame['Visa Class'] == 'H1B']

h1bNationalityDataFrame.to_excel('H1B_Files/H1B_By_Nationality_Q4_FY2023.xlsx', index=False, engine='openpyxl')

h1bNationalityDataFrame

Nationality Visa Class  Issuances      Month  Year
4              Afghanistan        H1B        5.0       JULY  2023
19                 Albania        H1B        8.0       JULY  2023
41                 Algeria        H1B        1.0       JULY  2023
68                  Angola        H1B        3.0       JULY  2023
82     Antigua and Barbuda        H1B        2.0       JULY  2023
...                    ...        ...        ...        ...   ...
11268           Uzbekistan        H1B        2.0  SEPTEMBER  2023
11297            Venezuela        H1B       19.0  SEPTEMBER  2023
11330              Vietnam        H1B       34.0  SEPTEMBER  2023
11366               Zambia        H1B        2.0  SEPTEMBER  2023
11384             Zimbabwe        H1B        6.0  SEPTEMBER  2023

[425 rows x 5 columns]

# Issuances by Nationality
Number of issuances for each Nationality for the whole quarter of the fiscal year

In [17]:
issuanceByNationality = h1bNationalityDataFrame.groupby('Nationality')['Issuances'].sum().reset_index()

issuanceByNationality.to_excel('H1B_Files/H1B_By_Nationality_totals_Q4_FY2023.xlsx', index=False, engine='openpyxl')

issuanceByNationality.sort_values(by=['Issuances'])

Nationality  Issuances
20             Bhutan        1.0
141       South Sudan        1.0
139           Somalia        1.0
152              Togo        1.0
35     Cayman Islands        1.0
..                ...        ...
103            Mexico      494.0
84       Korea, South      578.0
123       Philippines      764.0
37   China - mainland     7077.0
72              India    36001.0

[166 rows x 2 columns]

# Issuances by Post
Number of issuances for each Post for the whole quarter of the fiscal year

In [18]:
issuanceByPost = h1bPostDataFrame.groupby('Post')['Issuances'].sum().reset_index()

issuanceByPost.to_excel('H1B_Files/H1B_By_Post_totals_Q4_FY2023.xlsx', index=False, engine='openpyxl')

issuanceByPost.sort_values(by=['Issuances'])

Post  Issuances
93     Kolonia          1
117    Mbabane          1
172       Suva          1
127       Naha          1
111     Majuro          1
..         ...        ...
21     Beijing       2349
79   Hyderabad       6512
124     Mumbai       7078
131  New Delhi       8579
43     Chennai      11109

[196 rows x 2 columns]

# Issuance by Month (Post)
This gets the total issuance count for each month based on the post

In [20]:
issuanceByMonthPost = h1bPostDataFrame.groupby('Month')['Issuances'].sum().reset_index()

issuanceByMonthPost.to_excel('H1B_Files/H1B_Monthly_Totals_FY2023.xlsx', index=False, engine='openpyxl')

issuanceByMonthPost.sort_values(by=['Issuances'])

Month  Issuances
2  SEPTEMBER      14592
0       JULY      16564
1       JUNE      20808

# Issuance by Month (Nationality)
This gets the total issuance count for each month based on the post

In [11]:
issuanceByMonthNationality = h1bNationalityDataFrame.groupby('Month')['Issuances'].sum().reset_index()

issuanceByMonthNationality.sort_values(by=['Issuances'])

Month  Issuances
2  SEPTEMBER    14592.0
0       JULY    16564.0
1       JUNE    20808.0

# Average Issuance by Nationality
Looks at the average issuance by Nationality over Q4 of FY2023

In [21]:
averageByNationality = h1bNationalityDataFrame.groupby('Nationality')['Issuances'].mean().reset_index()

averageByNationality.to_excel('H1B_Files/H1B_by_Nationality_Average_FY2023.xlsx', index=False, engine='openpyxl')

averageByNationality.sort_values(by=['Issuances'])

Nationality     Issuances
52            Estonia      1.000000
141       South Sudan      1.000000
86             Kuwait      1.000000
20             Bhutan      1.000000
19            Bermuda      1.000000
..                ...           ...
103            Mexico    164.666667
84       Korea, South    192.666667
123       Philippines    254.666667
37   China - mainland   2359.000000
72              India  12000.333333

[166 rows x 2 columns]

# Average Issuance by Post
Looks at the average issuance by Post over Q4 of FY2023

In [22]:
averageByPost = h1bPostDataFrame.groupby('Post')['Issuances'].mean().reset_index()

averageByPost.to_excel('H1B_Files/H1B_by_Post_Average_FY2023.xlsx', index=False, engine='openpyxl')

averageByPost.sort_values(by=['Issuances'])

Post    Issuances
64      Freetown     1.000000
32   Brazzaville     1.000000
172         Suva     1.000000
174      Tallinn     1.000000
117      Mbabane     1.000000
..           ...          ...
21       Beijing   783.000000
79     Hyderabad  2170.666667
124       Mumbai  2359.333333
131    New Delhi  2859.666667
43       Chennai  3703.000000

[196 rows x 2 columns]

# Deploying to MongoDB

## Connecting to MongoDB

In [1]:
import pandas as pd
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi

uri = "mongodb+srv://JaydenLintz:0s9CcpobRMFYPbku@cluster0.lxnysiw.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0&tls=true&tlsVersion=TLS1_2"

client = MongoClient(uri, server_api=ServerApi('1'), tlsAllowInvalidCertificates=True)

# Send a ping to confirm a successful connection
try:
    client.admin.command('ping')
    print("Pinged your deployment. You successfully connected to MongoDB!")
except Exception as e:
    print(e)

C:\Users\lintz\AppData\Local\Programs\Python\Python310\lib\site-packages\pymongo\common.py:813: UserWarning: Unknown option tlsversion
  warnings.warn(str(exc))


SSL handshake failed: ac-xgprqzd-shard-00-01.lxnysiw.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:997),SSL handshake failed: ac-xgprqzd-shard-00-02.lxnysiw.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:997),SSL handshake failed: ac-xgprqzd-shard-00-00.lxnysiw.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:997), Timeout: 30s, Topology Description: <TopologyDescription id: 66171f3945228a3566e4f288, topology_type: ReplicaSetNoPrimary, servers: [<ServerDescription ('ac-xgprqzd-shard-00-00.lxnysiw.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: ac-xgprqzd-shard-00-00.lxnysiw.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:997)')>, <ServerDescription ('ac-xgprqzd-shard-00-01.lxnysiw.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: ac

# Analysis

## Issuance By Nationality
* Highest = India at 36001
* Lowest = Bhutan, South Sudan, Somalia, Togo, Cayman Islands... at 1

## Issuance By Post
* Highest = Chennai at 11109
* Lowest = Kolonia, Mbabane, Suva, Naha, Majuro... at 1

## Issuance by Month
* Highest = June at 20808
* Lowest = September at 14592

## Average Issuance by Nationality
* Highest = India at 12000.333
* Lowest = Estonia, South Sudan, Kuwait, Bhutan, Bermuda... at 1

## Average Issuance by Post
* Highest = Chennai at 3703
* Lowest = Freetown, Brazzaville, Suva, Tallinn, Mbabane... at 1
 Mbabane... at 1
	